// sql-client/lib/sqlite3-client.js
// Generated by CoffeeScript 1.10.0
(function() {
  var ConnectionFactory, HOMEDIR, LIB_COV, LIB_DIR, SQLClient, SQLClientPool, SQLite3Client, SQLite3ClientPool, SQLite3ConnectionFactory, fs, path, sqlite3,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty,
    slice = [].slice;

  fs = require('fs');

  path = require('path');

  HOMEDIR = path.join(__dirname, '..');

  LIB_COV = path.join(HOMEDIR, 'lib-cov');

  LIB_DIR = fs.existsSync(LIB_COV) ? LIB_COV : path.join(HOMEDIR, 'lib');

  SQLClient = require(path.join(LIB_DIR, 'sql-client')).SQLClient;

  SQLClientPool = require(path.join(LIB_DIR, 'sql-client-pool')).SQLClientPool;

  ConnectionFactory = require(path.join(LIB_DIR, 'connection-factory')).ConnectionFactory;

  sqlite3 = require('sqlite3').verbose();

  SQLite3ConnectionFactory = (function(superClass) {
    extend(SQLite3ConnectionFactory, superClass);

    function SQLite3ConnectionFactory() {
      this.execute = bind(this.execute, this);
      this.open_connection = bind(this.open_connection, this);
      return SQLite3ConnectionFactory.__super__.constructor.apply(this, arguments);
    }

    SQLite3ConnectionFactory.prototype.open_connection = function(filename, mode, callback) {
      var cb, db;
      if (typeof mode === 'function' && (callback == null)) {
        callback = mode;
        mode = null;
      }
      db = null;
      cb = (function(_this) {
        return function(err) {
          if (err != null) {
            return callback(err);
          } else {
            return callback(null, db);
          }
        };
      })(this);
      if (mode != null) {
        return db = new sqlite3.Database(filename, mode, cb);
      } else {
        return db = new sqlite3.Database(filename, cb);
      }
    };

    SQLite3ConnectionFactory.prototype.execute = function(db, sql, bindvars, callback) {
      if (bindvars == null) {
        bindvars = [];
      }
      return db.all(sql, bindvars, callback);
    };

    return SQLite3ConnectionFactory;

  })(ConnectionFactory);

  SQLite3Client = (function(superClass) {
    extend(SQLite3Client, superClass);

    function SQLite3Client() {
      var options;
      options = 1 <= arguments.length ? slice.call(arguments, 0) : [];
      SQLite3Client.__super__.constructor.apply(this, slice.call(options).concat([new SQLite3ConnectionFactory()]));
    }

    return SQLite3Client;

  })(SQLClient);

  SQLite3ClientPool = (function(superClass) {
    extend(SQLite3ClientPool, superClass);

    function SQLite3ClientPool() {
      var options;
      options = 1 <= arguments.length ? slice.call(arguments, 0) : [];
      SQLite3ClientPool.__super__.constructor.apply(this, slice.call(options).concat([new SQLite3ConnectionFactory()]));
    }

    return SQLite3ClientPool;

  })(SQLClientPool);

  exports.SQLite3ConnectionFactory = SQLite3ConnectionFactory;

  exports.SQLite3Client = SQLite3Client;

  exports.SQLite3ClientPool = SQLite3ClientPool;

}).call(this);
