// sql-client/lib/bin/mysql-runner.js
// Generated by CoffeeScript 1.10.0
(function() {
  var HOMEDIR, LIB_COV, LIB_DIR, MySQLClient, MySQLRunner, ObjectUtil, SQLRunner, fs, path,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty,
    slice = [].slice;

  fs = require('fs');

  path = require('path');

  HOMEDIR = path.join(__dirname, '..', '..');

  LIB_COV = path.join(HOMEDIR, 'lib-cov');

  LIB_DIR = fs.existsSync(LIB_COV) ? LIB_COV : path.join(HOMEDIR, 'lib');

  SQLRunner = require(path.join(LIB_DIR, 'sql-runner')).SQLRunner;

  MySQLClient = require(path.join(LIB_DIR, 'mysql-client')).MySQLClient;

  ObjectUtil = require('inote-util').ObjectUtil;

  MySQLRunner = (function(superClass) {
    extend(MySQLRunner, superClass);

    function MySQLRunner(connect_string, options) {
      this._stringify_results = bind(this._stringify_results, this);
      this._handle_argv = bind(this._handle_argv, this);
      this._get_options = bind(this._get_options, this);
      this.set_client = bind(this.set_client, this);
      var client;
      if ((connect_string != null) && typeof connect_string === 'object' && (options == null)) {
        options = connect_string;
        connect_string = null;
      }
      client = null;
      if (connect_string != null) {
        client = new MySQLCLient(connect_string);
      }
      MySQLRunner.__super__.constructor.call(this, client, options);
    }

    MySQLRunner.prototype.set_client = function(client) {
      if (client.execute == null) {
        client = new MySQLClient(client);
      }
      return MySQLRunner.__super__.set_client.call(this, client);
    };

    MySQLRunner.prototype._get_options = function(additional) {
      var my_opts;
      if (additional == null) {
        additional = {};
      }
      my_opts = {
        d: {
          alias: 'db',
          describe: "Databse connect string."
        }
      };
      return MySQLRunner.__super__._get_options.call(this, ObjectUtil.merge(my_opts, additional));
    };

    MySQLRunner.prototype._handle_argv = function(argv) {
      if (argv.db != null) {
        this.set_client(argv.db);
      }
      return MySQLRunner.__super__._handle_argv.call(this, argv);
    };

    MySQLRunner.prototype._stringify_results = function() {
      var fields, other, rows;
      rows = arguments[0], fields = arguments[1], other = 3 <= arguments.length ? slice.call(arguments, 2) : [];
      if (rows != null) {
        return JSON.stringify(rows, null, 2);
      } else {
        return MySQLRunner.__super__._stringify_results.apply(this, [rows, fields].concat(slice.call(other)));
      }
    };

    return MySQLRunner;

  })(SQLRunner);

  exports.MySQLRunner = MySQLRunner;

  if (require.main === module) {
    (new MySQLRunner()).main();
  }

}).call(this);
